android/src/com/mapswithme/maps/widget/recycler/GridDividerItemDecoration.java

```diff
@@ -0,0 +1,133 @@
+   * @param parent The RecyclerView onto which dividers are being added
+   * @param state  The current RecyclerView.State of the RecyclerView
+   */
+  @Override
+  public void onDraw(Canvas canvas, RecyclerView parent, RecyclerView.State state)
+  {
+    drawHorizontalDividers(canvas, parent);
+    drawVerticalDividers(canvas, parent);
+  }
+
+  /**
+   * Determines the size and location of offsets between items in the parent
+   * RecyclerView.
+   *
+   * @param outRect The {@link Rect} of offsets to be added around the child view
+   * @param view    The child view to be decorated with an offset
+   * @param parent  The RecyclerView onto which dividers are being added
+   * @param state   The current RecyclerView.State of the RecyclerView
+   */
+  @Override
+  public void getItemOffsets(Rect outRect, View view, RecyclerView parent, RecyclerView.State state)
+  {
+    super.getItemOffsets(outRect, view, parent, state);
+
+    boolean childIsInLeftmostColumn = (parent.getChildAdapterPosition(view) % mNumColumns) == 0;
+    if (!childIsInLeftmostColumn)
+    {
+      outRect.left = mHorizontalDivider.getIntrinsicWidth();
+    }
+
+    boolean childIsInFirstRow = (parent.getChildAdapterPosition(view)) < mNumColumns;
+    if (!childIsInFirstRow)
+    {
+      outRect.top = mVerticalDivider.getIntrinsicHeight();
+    }
+  }
+
+  /**
+   * Adds horizontal dividers to a RecyclerView with a GridLayoutManager or
+   * its subclass.
+   *
+   * @param canvas The {@link Canvas} onto which dividers will be drawn
+   * @param parent The RecyclerView onto which dividers are being added
+   */
+  private void drawHorizontalDividers(Canvas canvas, RecyclerView parent)
+  {
+    int parentTop = parent.getPaddingTop();
+    int parentBottom = parent.getHeight() - parent.getPaddingBottom();
+
+    for (int i = 0; i < mNumColumns; i++)
+    {
+      View child = parent.getChildAt(i);
+      RecyclerView.LayoutParams params = (RecyclerView.LayoutParams) child.getLayoutParams();
+
+      int parentLeft = child.getRight() + params.rightMargin;
+      int parentRight = parentLeft + mHorizontalDivider.getIntrinsicWidth();
+
+      mHorizontalDivider.setBounds(parentLeft, parentTop, parentRight, parentBottom);
+      mHorizontalDivider.draw(canvas);
+    }
+  }
+
+  /**
+   * Adds vertical dividers to a RecyclerView with a GridLayoutManager or its
+   * subclass.
+   *
+   * @param canvas The {@link Canvas} onto which dividers will be drawn
+   * @param parent The RecyclerView onto which dividers are being added
+   */
+  private void drawVerticalDividers(Canvas canvas, RecyclerView parent)
+  {
+    int parentLeft = parent.getPaddingLeft();
+    int parentRight = parent.getWidth() - parent.getPaddingRight();
+
+    int childCount = parent.getChildCount();
+    int numChildrenOnLastRow = childCount % mNumColumns;
+    int numRows = childCount / mNumColumns;
+    if (numChildrenOnLastRow == 0)
+    { // TODO: Replace this with math
```

Comment:
Did you mean ```numRows = (childCount - 1) / mNumColumns;```?
Do you really want numRows to be zero when childCount is not greater than mNumColumns? The name seems misleading.
I suggest that you either rename it to numDividers or set it to the proper value and iterate up to ```numRows - 1```.